server/src/main/java/org/opensearch/indices/replication/common/ReplicationTarget.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.indices.replication.common;
+
+import org.apache.logging.log4j.Logger;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.ActionListener;
+import org.opensearch.common.UUIDs;
+import org.opensearch.common.bytes.BytesReference;
+import org.opensearch.common.logging.Loggers;
+import org.opensearch.common.util.concurrent.AbstractRefCounted;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.index.store.Store;
+import org.opensearch.index.store.StoreFileMetadata;
+import org.opensearch.indices.recovery.MultiFileWriter;
+import org.opensearch.indices.recovery.RecoveryRequestTracker;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public abstract class ReplicationTarget extends AbstractRefCounted {
+
+    private static final AtomicLong ID_GENERATOR = new AtomicLong();
+
+    // last time the target/status was accessed
+    private volatile long lastAccessTime = System.nanoTime();
+    private final RecoveryRequestTracker requestTracker = new RecoveryRequestTracker();
+    private final long id;
+
+    protected final AtomicBoolean finished = new AtomicBoolean();
+    protected final IndexShard indexShard;
+    protected final Store store;
+    protected final ReplicationListener listener;
+    protected final MultiFileWriter multiFileWriter;
+    protected final Logger logger;
+    protected final ReplicationLuceneIndex recoveryStateIndex;
+
+    protected abstract String getPrefix();
+
+    protected abstract void onDone();
+
+    protected abstract void onCancel(String reason);
+
+    protected abstract void onFail(OpenSearchException e, boolean sendShardFailure);
+
+    public abstract ReplicationState state();
+
+    public ReplicationTarget(String name, IndexShard indexShard, ReplicationLuceneIndex recoveryStateIndex, ReplicationListener listener) {
+        super(name);
+        this.logger = Loggers.getLogger(getClass(), indexShard.shardId());
+        this.listener = listener;
+        this.id = ID_GENERATOR.incrementAndGet();
+        this.recoveryStateIndex = recoveryStateIndex;
+        this.indexShard = indexShard;
+        this.store = indexShard.store();
+        final String tempFilePrefix = getPrefix() + UUIDs.randomBase64UUID() + ".";
+        this.multiFileWriter = new MultiFileWriter(indexShard.store(), recoveryStateIndex, tempFilePrefix, logger, this::ensureRefCount);
+        // make sure the store is not released until we are done.
+        store.incRef();
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    /**
+     * return the last time this RecoveryStatus was used (based on System.nanoTime()
```

Comment:
Nit - There are several mentions of "recovery" in this file which should appropriately replaced by "replication"